Draft commit (must be splitted)

aderiabin opened the pull request.

aderiabin committed:
Add and remove whitespaces

aderiabin committed:
Code readability improvement

aderiabin committed:
Improve ATF logger with binary data size information

aderiabin committed:
Fix creation and removing of ATF event expectations

aderiabin committed:
Security manager component implementation

aderiabin committed:
Secure communication functionality implementation

aderiabin committed:
Frame level functionality implementation

aderiabin committed:
Update ATF config with Security related settings

aderiabin committed:
Add LDoc documentation for security functionality

aderiabin committed:
Fixes according review comments (Lua part)

aderiabin committed:
Fixes according review comments (Lua part)

aderiabin committed:
Refactoring according review comment.

aderiabin committed:
New security constants were added.

aderiabin committed:
Fixes according review comments (C part)

aderiabin committed:
Fix for heartbeat and security conflict issue.
Update for ATF logger - added log for contril messages (Start session,
End session, heartbeat)

aderiabin committed:
Added implementation of client certificate validation

aderiabin committed:
Secure streaming implementation

LuxoftAKutsan committed:
Use CMake for ATF build

LuxoftAKutsan committed:
Add ATF HMI adapter for MqTCP

Commit:
Remote lua client and remote mq server implementation

AByzhynar committed:
Initial commit

dboltovskyi committed:
Add support of V5 SDL protocol

AByzhynar committed:
Fixes part2

aderiabin committed:
Build lua_xml library as subproject

aderiabin committed:
Update CMakeLists with install instructions and minimal cmake version

aderiabin committed:
Update start.sh with new executable name

aderiabin committed:
Add build folder into git ignore list

AByzhynar committed:
Add possibility to create parametrized mqueue

aderiabin committed:
Draft commit (must be splitted)

LuxoftAKutsan commented on modules/config.lua:
```diff
@@ -9,10 +9,20 @@
-config.hmiPort = 8087
+
+--- HMI configuration
+config.hmiAdapterType = "WebSocket"
```
@aderiabin move hmiAdapterType to hmiAdapterConfig

aderiabin committed:
remote_adapter_client/CMakeLists.txt was updated for Qt related build

aderiabin committed:
Remove empty source lua_lib/lua_sdl_remote_adapter.cc

aderiabin committed:
Add Qt-based implementation for lua_sdl_remote_adapter_library

aderiabin committed:
Update lua implementation for remote_hmi_adapter

aderiabin committed:
Minor update for base sdl_remote_adapter_client

BSolonenko commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_receive_thread.cc:
```diff
@@ -0,0 +1,31 @@
+        }
+}
+
+void SDLRemoteTestAdapterReceiveThread::run() {
```
@aderiabin Maybe you like this decision more?
```
QString receivedData(client_->receive().first.c_str());

while(!receivedData.empty()){
    emit dataAvailable(receivedData);
    receivedData = client_->receive().first.c_str();
}

emit sourceUnreachable();
```

aderiabin replied:
@BSolonenko Applied in 716e1c1

BSolonenko commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.cc:
```diff
@@ -0,0 +1,125 @@
+                                             MqParams& in_params,
+                                             MqParams& out_params) {
+    try {
+        remote_adapter_client_ptr_ = new SDLRemoteTestAdapterClient(host, port);
```
@aderiabin  Why do you not clean memory? If this method is called twice, you will lose one pointer.

aderiabin replied:
@BSolonenko Fixed in 716e1c1.
Additionaly all connection data is stored in SDLRemoteTestAdapterQtClient now.

dboltovskyi committed:
Reduce count of global and unused variables in ATF

dboltovskyi committed:
Fix for heartbeat

dboltovskyi committed:
Change default security protocol to DTLS

aderiabin committed:
Fix for potential memory leaks

BSolonenko commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.cc:
```diff
@@ -19,56 +28,35 @@ SDLRemoteTestAdapterQtClient::~SDLRemoteTestAdapterQtClient() {
+            remote_adapter_client_ptr_ = new SDLRemoteTestAdapterClient(host_, port_);
+            int in_open_result = openWithParams(in_mq_params_);
+            int out_open_result = openWithParams(out_mq_params_);
+            if ((0 == in_open_result) && (0 == out_open_result)) {
```
@aderiabin  Maybe the situation when you get into this state twice? If so, you may still have a memory leak.

SNiukalov replied:
@aderiabin We can use shared or unique pointer
instead of simple pointers ?

EKuliiev replied:
@aderiabin use std::unique_ptr for remote_adapter_client_ptr_ and for listener_ptr_

aderiabin replied:
@SNiukalov @EKuliiev @BSolonenko `std::unique_ptr` is used for pointers on SDLRemoteTestAdapterClient and SDLRemoteTestAdapterReceiveThread instances 
(see bc3f86577b740bc84201d7d0edaf82cce1d8ec52)

SNiukalov commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.cc:
```diff
@@ -0,0 +1,90 @@
+            int in_open_result = openWithParams(in_mq_params_);
+            int out_open_result = openWithParams(out_mq_params_);
+            if ((0 == in_open_result) && (0 == out_open_result)) {
+                listener_ptr_ = new SDLRemoteTestAdapterReceiveThread(this);
```
We can use shared or unique pointer
instead of simple pointers ?

aderiabin replied:
@SNiukalov `std::unique_ptr` is used for pointers on SDLRemoteTestAdapterClient and SDLRemoteTestAdapterReceiveThread instances 
(see bc3f86577b740bc84201d7d0edaf82cce1d8ec52)

SNiukalov commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.h:
```diff
@@ -0,0 +1,111 @@
+    MqParams in_mq_params_;
+    MqParams out_mq_params_;
+    SDLRemoteTestAdapterClient* remote_adapter_client_ptr_ = nullptr;
+    SDLRemoteTestAdapterReceiveThread* listener_ptr_ = nullptr;
```
We can use shared or unique pointer
instead of simple pointers ?

aderiabin replied:
@SNiukalov `std::unique_ptr` is used for pointers on SDLRemoteTestAdapterClient and SDLRemoteTestAdapterReceiveThread instances 
(see bc3f86577b740bc84201d7d0edaf82cce1d8ec52)

EKuliiev commented on src/remote_adapter/remote_adapter_client/CMakeLists.txt:
```diff
@@ -1,15 +1,52 @@
+find_package(Qt5 COMPONENTS Core REQUIRED)
 find_package( Threads REQUIRED )
 FIND_PACKAGE(Lua 5.2 EXACT REQUIRED)
-message(STATUS "Lua found : " ${LUA_INCLUDE_DIR})
```
@aderiabin  please change FIND_PACKAGE to find_package

aderiabin replied:
@EKuliiev Applied (see 497ec85ac829fefb39ccf5198d0ad6ac798e7b74)

EKuliiev commented on src/remote_adapter/remote_adapter_client/CMakeLists.txt:
```diff
@@ -1,15 +1,52 @@
 FIND_PACKAGE(Lua 5.2 EXACT REQUIRED)
-message(STATUS "Lua found : " ${LUA_INCLUDE_DIR})
+
+# list(GET LUA_LIBRARIES 0 LUA_LIB)
```
@aderiabin please remove this commented part( lines 29..42) since lua imported library already provided in root cmake list

aderiabin replied:
@EKuliiev Current part of code was moved into if condition, because it is needed for separate compilation of module `remote_adapter_client`
(see 497ec85ac829fefb39ccf5198d0ad6ac798e7b74)

Jack-Byrne committed:
Merge pull request #136 from smartdevicelink/fix/refactoring_and_fixes

Security ATF feature prerequisites (Refactoring and fixes)

Jack-Byrne committed:
Merge pull request #107 from smartdevicelink/feature/security

Security ATF feature

EKuliiev commented on src/remote_adapter/remote_adapter_client/lua_lib/sdl_remote_adapter_client_lua_wrapper.h:
```diff
@@ -1,8 +1,8 @@
-#include "lua.h"
-#include "lualib.h"
-#include "lauxlib.h"
+#include <lua5.2/lua.h>
```
@aderiabin I think better to avoid such prefix lua5.2 before include
see recommendations:
https://cmake.org/cmake/help/v3.6/module/FindLua.html

If no appropriate version found cmake fails with error because of find_package(Lua 5.2 REQAIRED)  

use just 
`#include <lua.h> 

In you case it works just because liblua installed in default path.

aderiabin replied:
@EKuliiev Applied (see 16996cc5aa4b9fa3938a0dd47c37e1f8172677a3)

EKuliiev commented on src/remote_adapter/remote_adapter_client/qt_impl/lua_lib/sdl_remote_adapter_qt_client_lua_wrapper.cc:
```diff
@@ -0,0 +1,213 @@
+  static const luaL_Reg SDLRemoteTestAdapterFunctions[] = {
+      {"connect", SDLRemoteTestAdapterLuaWrapper::lua_connect},
+      {"write", SDLRemoteTestAdapterLuaWrapper::lua_write},
+
```
@aderiabin redundant new line

aderiabin replied:
@KVGrygoriev It will be applied as part of future refactoring.

aderiabin replied:
@EKuliiev Applied (see 1c6f54a5a5c135f34813d35ec8dc5521604899f0)

EKuliiev commented on src/remote_adapter/remote_adapter_client/qt_impl/lua_lib/sdl_remote_adapter_qt_client_lua_wrapper.h:
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+extern "C" {
+#include <lua5.2/lua.h>
```
@aderiabin #include <lua5.2/lua.h> --> #include <lua.h>

aderiabin replied:
@EKuliiev Applied (see 16996cc5aa4b9fa3938a0dd47c37e1f8172677a3)

EKuliiev commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.h:
```diff
@@ -0,0 +1,111 @@
+    // * @param host server host to connect
+    // * @param port server port to connect
+    // */
+    // void connect(const std::string& host, uint32_t port, const std::string& name);
```
@aderiabin if this is not used please delete commented code

aderiabin replied:
@EKuliiev @KVGrygoriev  Unused code was removed (see bc3f86577b740bc84201d7d0edaf82cce1d8ec52)

EKuliiev commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.h:
```diff
@@ -0,0 +1,111 @@
+#include <utility>
+
+#include "sdl_remote_adapter_client.h"
+// #include "qt_impl/sdl_remote_adapter_receive_thread.h"
```
@aderiabin if this include is redundant please remove

aderiabin replied:
@EKuliiev Unused code was removed (see bc3f86577b740bc84201d7d0edaf82cce1d8ec52)

EKuliiev reviewed (commented):
see review issues

KVGrygoriev commented on modules/connecttest.lua:
```diff
@@ -26,6 +26,13 @@ local SDL = require('SDL')
+  hmiAdapter = require('remote_hmi_adapter_connection')
+else if config.hmiAdapterType == "WebSocket" then
+  hmiAdapter = require('websocket_connection')
+end
```
would be nice notify user in case no one variant match (typo or etc.)

aderiabin replied:
@KVGrygoriev Wrong HMI adapter handling was added (see f406fd26b2ba8306caa87ead9f83a9133efd7ca2)

KVGrygoriev commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.cc:
```diff
@@ -0,0 +1,90 @@
+
+SDLRemoteTestAdapterQtClient::~SDLRemoteTestAdapterQtClient() {
+    if (nullptr != listener_ptr_) {
+        delete(listener_ptr_);
```
@aderiabin, make it easier, just 
```if (listener_ptr_)```

aderiabin replied:
@KVGrygoriev Not actual already, code was removed.

KVGrygoriev commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.cc:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    if (nullptr != remote_adapter_client_ptr_) {
+        remote_adapter_client_ptr_->clear();
```
the same like above

aderiabin replied:
@KVGrygoriev It will be applied as part of future refactoring.

aderiabin replied:
@KVGrygoriev Applied (see 1c6f54a5a5c135f34813d35ec8dc5521604899f0)

KVGrygoriev commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.cc:
```diff
@@ -0,0 +1,90 @@
+    if (0 == result) {
+        emit bytesWritten(data.length());
+    }
+    else if (1 == result) {
```
Would be nice to make some wrapper for 0 and 1 values, like SUCCESS & FAILED

aderiabin replied:
@KVGrygoriev Wrapper is used (see bc3f86577b740bc84201d7d0edaf82cce1d8ec52)

KVGrygoriev commented on src/remote_adapter/remote_adapter_client/sdl_remote_adapter_client.h:
```diff
@@ -73,7 +73,7 @@ class SDLRemoteTestAdapterClient {
  private:
-  std::string data_;
+  // std::string data_;
   rpc::client connection_;
```
remove commented code, please

KVGrygoriev commented on src/remote_adapter/remote_adapter_client/sdl_remote_adapter_client.h:
```diff
@@ -73,7 +73,7 @@ class SDLRemoteTestAdapterClient {
 
  private:
-  std::string data_;
+  // std::string data_;
```
remove commented code, please

aderiabin replied:
@KVGrygoriev Removed (see 79905c73bd7d828a0b6dc6b482f1cc9ec901cc1d)

dboltovskyi committed:
Add default value for 'bandwidth' parameter

dboltovskyi committed:
Merge pull request #137 from dboltovskyi/fix/add_default_value_for_bandwidth

Default value for 'bandwidth' parameter

aderiabin committed:
Add notification about wrong HMI adapter name. Fix QT signal connection for RemoteHMIAdapter in lua.

aderiabin committed:
Remove unused code

aderiabin committed:
Fix lua wrapper

aderiabin committed:
Fix remote_adapter_qt_client

aderiabin committed:
Fix CMakeLists.txt

aderiabin commented:
@BSolonenko @SNiukalov @EKuliiev @KVGrygoriev Thank you for your review! Please find my changes according your comments above.

Also you can find fixes for couple major issues here:
 - Wrong Lua stack use issue
 - ATF Segmentation fault

aderiabin committed:
Run remote_adapter_server in async mode

aderiabin committed:
Refactor lua wrapper for remote_adapter_client

EKuliiev commented on CMakeLists.txt:
```diff
@@ -27,6 +27,7 @@ set_target_properties(lua::lua
     IMPORTED_LINK_DEPENDENT_LIBRARIES ${LIBM_LIB}
     )
 
+set(LUA_FOUND TRUE)
```
@aderiabin set(LUA_FOUND TRUE) is redundant.
find_package provides var <package>_FOUND that in your case is Lua_FOUND
https://cmake.org/cmake/help/v3.8/command/find_package.html?highlight=i

EKuliiev commented on src/remote_adapter/remote_adapter_client/CMakeLists.txt:
```diff
@@ -1,15 +1,54 @@
-FIND_PACKAGE(Lua 5.2 EXACT REQUIRED)
-message(STATUS "Lua found : " ${LUA_INCLUDE_DIR})
+
+if(NOT LUA_FOUND)
```
@aderiabin use Lua_FOUND provided be FindLua.cmake script.
I've note it above.

LuxoftAKutsan commented on CMakeLists.txt:
```diff
@@ -27,6 +27,7 @@ set_target_properties(lua::lua
     IMPORTED_LINK_DEPENDENT_LIBRARIES ${LIBM_LIB}
     )
 
+set(LUA_FOUND TRUE)
```
@aderiabin that if lua is not on the system, will still be true?

aderiabin replied:
@LuxoftAKutsan In case appropriate Lua version is not found in system cmake will be failed, so LUA_FOUND will be set only in case of successful Lua version find.

This flag was added for support separate building of subproject `remote_adapter`
I am ready to accept any better solution for this issue.

EKuliiev replied:
@aderiabin Try next:
```
if (NOT TARGET lua::lua)
add_library(lua::lua ....) # create imported target
# set target props you need
endif()
```

aderiabin replied:
@EKuliiev Applied (see c09eab5)
Thank you.

LuxoftAKutsan commented on src/remote_adapter/remote_adapter_client/qt_impl/lua_lib/sdl_remote_adapter_qt_client_lua_wrapper.cc:
```diff
@@ -0,0 +1,199 @@
+  // Index -4 - string host
+  // index -5 - Library table
+
+  auto out_mq_params = build_MqParams(L);
```
@aderiabin may be function build_MqParams will remove value from the top of the stack and used in lua library

aderiabin replied:
@LuxoftAKutsan It is not principal. Current decision was chosen to simplify understanding of Lua stack manipulations and unify Lua stack manipulation approach for simple and structured types of data (get data + remove data from stack ) :
```
  //Simple type
  // Get data from Lua stack
  auto port = lua_tointeger(L, -1); 
  auto ip = lua_tostring(L, -2);
  // Remove 2 values from top of the stack
  lua_pop(L, 2);  
  
  // Structured type
  // Get data from Lua stack
  auto out_mq_params = build_MqParams(L);
  // Remove value from the top of the stack
  lua_pop(L, 1);  
```

LuxoftAKutsan commented on src/remote_adapter/remote_adapter_client/qt_impl/lua_lib/sdl_remote_adapter_qt_client_lua_wrapper.h:
```diff
@@ -0,0 +1,22 @@
+namespace lua_lib {
+
+struct SDLRemoteTestAdapterLuaWrapper {
+  static int create_SDLRemoteTestAdapter(lua_State* L);
```
@aderiabin functions require doxygen description a d splitting to public/private

LuxoftAKutsan commented on src/remote_adapter/remote_adapter_client/qt_impl/sdl_remote_adapter_qt_client.cc:
```diff
@@ -0,0 +1,86 @@
+}
+
+void SDLRemoteTestAdapterQtClient::connectMq() {
+    if (!isconnected_) {
```
@aderiabin I would invert this if to avoid deep nested blocks

aderiabin replied:
@LuxoftAKutsan Applied (see fd8a7a5 )

aderiabin committed:
Add ability to send control messages through control MQ to remote_hmi_adapter

aderiabin committed:
Add control method

aderiabin committed:
Fix lua::lua target adding issue

aderiabin committed:
Add console messages for remote server and performed not significant refactoring

aderiabin committed:
Add console messages for remote client and performed not significant refactoring

aderiabin committed:
Fix install destination for c++ modules

aderiabin committed:
Fix cmake issue with independent build of RemoteTestingAdapterServer

aderiabin committed:
Add functions close and unlink to SDLRemoteTestAdapterClient. Fix console reporting.

aderiabin committed:
Implement separate modules Remote HMI adapter and Remote MQ handler on base on single RPC connection

aderiabin committed:
Add array wrapping for HMI messages for Sync platform support

SNiukalov committed:
Trivial fix for connection.

SNiukalov committed:
Added shared memory manager.

SNiukalov committed:
Adjustment of mqueue manager components.

dboltovskyi committed:
Fix calculation of multiframe message size

dboltovskyi committed:
Merge pull request #138 from dboltovskyi/fix/fix_calculation_of_multi_frame_message_size

Fix calculation of multi-frame message size

dboltovskyi committed:
Fix handling of events when SDL stopped

SNiukalov committed:
Fixed flags, that control the operation(Read,Write) in message queue.
Fixed mobile Port on 12346 (default 12345 is busy tokenmanager).

SNiukalov committed:
Added temporary code to the start & stop sdl.

SNiukalov committed:
Cleaning of old messages in the queue is added.
Fixed reading of shared memory in MqReceive.
Added a detailed errors output for the queue.

SNiukalov committed:
Shared memory: added RPC to the remote server.

aderiabin committed:
Implement client part of remote HMI adapter with shared memory support

SNiukalov committed:
Updated constants.

SNiukalov committed:
Shared memory: updated functionality.

SNiukalov committed:
Added Utils manager.

dboltovskyi committed:
Merge pull request #139 from dboltovskyi/fix/handling_of_events_when_sdl_stopped

Fix handling of events when SDL stopped

SNiukalov committed:
UtilsManager: Updated base functionality.

aderiabin committed:
Implement client part of remote utils functionality

Commit:
applink connection module prorotype

Commit:
added applink_connection send message and changes in connecttest

dboltovskyi committed:
Update handling of SDL build options

dboltovskyi committed:
Merge pull request #140 from dboltovskyi/fix/update_handling_of_sdl_build_options

Update handling of sdl build options

aderiabin committed:
Update client part of remote utils functionality with result code for data receive functions

aderiabin committed:
Update lua part of remote utils functionality with result checks

Commit:
fxing indentation and remoed commented code

Commit:
added logic for Remote SDL start

Commit:
fixes in SDL and connecttest modules according to PR notes

Commit:
fix of error messages

SNiukalov committed:
Fixed build for Linux.

SNiukalov committed:
UtilsManager: Updated functionality for the files manipulation.

SNiukalov committed:
UtilsManager: Updated functionality for obtaining large data in file.

dboltovskyi committed:
Fix local run

dboltovskyi committed:
FixUp

SNiukalov committed:
Deleted start and stop "SDL" at server startup.

SNiukalov committed:
Fixed result code.

SNiukalov committed:
Added executing bash command.

dboltovskyi committed:
FixUp

SNiukalov committed:
Updated functionality for working with paths.

SNiukalov committed:
Added ExecuteCommand in lua.
Updated relation between wrapper & client.

dboltovskyi committed:
FixUp

dboltovskyi committed:
FixUp

dboltovskyi committed:
Replace CRASH status by IDLE for remote

dboltovskyi committed:
FixUp

dboltovskyi committed:
Enable reading from .ini

dboltovskyi committed:
FixUp

dboltovskyi committed:
Fix updating of path parameters

SNiukalov committed:
UtilsManager: Updated existing functionality for Restore/Backup file.

dboltovskyi committed:
Remove ATF binaries

dboltovskyi committed:
Increase timer for SDL logger

dboltovskyi committed:
Switch on SDL logger

dboltovskyi committed:
Implement single parameter for FolderExists and FolderDelete

dboltovskyi committed:
Add bash script runner

SNiukalov committed:
Fixed work with the shared memory.

SNiukalov committed:
Implement single parameter for CreateFolder.

SNiukalov committed:
Fixed FolderDelete functionality.

dboltovskyi committed:
Add specific path to SDL policy DB

SNiukalov committed:
Removed rpc response_error for mq_receive.
The error is returned in the result code.

SNiukalov committed:
Fixed work with shared memory.

SNiukalov committed:
Small optimization when writing to shared memory.

dboltovskyi committed:
Fix after rebase onto develop

SNiukalov committed:
Adjusting mq & shared memory for using existing channel.
Added handler for SIGSEGV.

dboltovskyi committed:
Remove not required print

SNiukalov committed:
Minor fixes for logging.

SNiukalov committed:
Adjusting mq and shared memory for proper operation after the script is abort.

dboltovskyi committed:
Fix for LuaOpenSSL lib

SNiukalov committed:
The creation of core dumps in the SIGSEGV handler is included.

SNiukalov committed:
The current logging method has been changed.
Added the ability to turn off logging when compiling.
Added logging for "rpclib".

SNiukalov committed:
Updated the output for the log.

SNiukalov committed:
Fixed checking the return code for the mq_open.

SNiukalov committed:
Small changes to check the return code.

SNiukalov committed:
Updated the output for the log.

dboltovskyi committed:
Change PolicyDB to it's configuration

SNiukalov committed:
Added start and stop "AppLinkService" & "SmartDeviceLink"
at server startup.

SNiukalov committed:
Fixed thread stop.

SNiukalov committed:
Added cleaning of the lua environment after aborting script.

Commit:
added fullAppID policy feature

Commit:
Merge branch 'feature/remote_atf_connection' of https://github.com/aderiabin/sdl_atf into feature/remote_atf_connection

dboltovskyi committed:
Update SDL module with new functions

dboltovskyi committed:
Change starting Id of requests from HMI

dboltovskyi committed:
Add support of certificates

dboltovskyi committed:
Update SDL module

dboltovskyi committed:
Additional improvements

Commit:
Improvements for restore and deletion files

aderiabin committed:
Update CMakeLists.txt

Add ./tools to directory copy list for `make install` command execution.

Commit:
updated according to changes in Open ATF

Commit:
Improve set method and cleaning storage

Commit:
Merge branch 'feature/remote_atf_connection' of https://github.com/aderiabin/sdl_atf into feature/remote_atf_connection
